### DISCUSSION TITLE
How to get jdk.ObjectAllocationOutsideTLAB events?

### DISCUSSION THREAD

**Discussion by jolivares** [async-profiler/async-profiler#958] [Q&A] [answered]
Running the following code generates `jdk.ObjectAllocationOutsideTLAB` events
```
///usr/bin/env jbang "$0" "$@" ; exit $?
//JAVA 21
//JAVA_OPTIONS -Xmx512m -XX:+UseG1GC -XX:StartFlightRecording=filename=profile.jfr,settings=tlab_profile.jfc

import java.util.ArrayList;

class G1HumongousObjects {
	public static void main(String[] args) {
		for (int i = 0; i < 10_000; i++) {
			var list = new ArrayList<String>();
			for (int j = 0; j < 10_000; j++) {
				list.add("item" + i + "-" + j);
			}
		}
	}
}
```
```
$ jabang G1HumongousObjects.java
[jbang] Building jar for G1HumongousObjects.java...
[0.243s][info][jfr,startup] Started recording 1. No limit specified, using maxsize=250MB as default.
[0.243s][info][jfr,startup] 
[0.243s][info][jfr,startup] Use jcmd 86222 JFR.dump name=1 to copy recording data to file.

$ jfr summary profile.jfr | grep TLAB
 jdk.ObjectAllocationInNewTLAB           19367        286342
 jdk.ObjectAllocationOutsideTLAB          2339         32304
 jdk.GCTLABConfiguration                     1            11
```

I've explicitly enabled `ObjectAllocationInNewTLAB` and `ObjectAllocationOutsideTLAB` in `tlab_profile.jfc`.

However, If I remove the `-XX:StartFlightRecording=filename=profile.jfr,settings=tlab_profile.jfc` part from the file, and run the file again using async profiler as agent
```
$ jbang --javaagent=ap-loader@jvm-profiling-tools/ap-loader=start,event=alloc,file=profile.jfr G1HumongousObjects.java
[jbang] Building jar for scratch.java...
WARNING: A JVM TI agent has been loaded dynamically (/Users/xxxxx/Library/Application Support/me.bechberger.ap-loader/3.0/lib/libasyncProfiler.dylib)
WARNING: If a serviceability tool is in use, please run with -XX:+EnableDynamicAgentLoading to hide this warning
WARNING: Dynamic loading of agents will be disallowed by default in a future release
Profiling started
```
I only get `jdk.ObjectAllocationInNewTLAB` events:
```
$ jfr summary profile.jfr | grep TLAB
 jdk.ObjectAllocationInNewTLAB       13854        263850
 jdk.ObjectAllocationOutsideTLAB         0             0
```
Is there any flag/config setting that allows me to get `jdk.ObjectAllocationOutsideTLAB` events using async-profiler?

**Comment by apangin** [comment, marked as answer]
Since JDK 11, async-profiler uses low overhead JVM TI object sampler for allocation profiling, which does not distinguish between allocations in TLAB and outside TLAB. Instead, it samples every `N` allocated kilobytes, regardless of where the object is allocated.

You may still enable legacy JFR events when running async-profiler. To do so, use `--jfrsync` option:
```
asprof -e cpu --jfrsync +jdk.ObjectAllocationInNewTLAB+jdk.ObjectAllocationOutsideTLAB -f profile.jfr PID
```
or, when running as an agent:
```
java -agentlib:asyncProfiler=start,event=cpu,jfrsync=+jdk.ObjectAllocationInNewTLAB+jdk.ObjectAllocationOutsideTLAB,file=profile.jfr ...
```

**Comment by shivongit** [comment]
Why are we going away from jfr way of profiling object allocation. According to https://bugs.openjdk.org/browse/JDK-8257602 there are issues with profiling based on newTLAB and outsideTLAB on jfr and changes based on throttling are suggested. 
Are there any advantages of JVM TI object sampler wrt changes suggested in post with respect to product stability, profiling.

**Reply by apangin** [reply]
As mentioned above, async-profiler does _not_ use `ObjectAllocationInNewTLAB`/`ObjectAllocationOutsideTLAB` events since JDK 11.

JVM TI sampler has low overhead, it provides precise configuration of sampling interval, it nicely integrates with other parts of async-profiler as it relies on native API, it gives async-profiler control on stacktrace collection mechanism: in particlar, allows collecting native frames, which JFR has never supported. JVM TI sampler also makes possible to customize live object profiling (#633).

That being said, async-profiler does not "go away" from JFR; you can still use all JFR capabilities with async-profiler, specificially, switch to `ObjectAllocationSample` event if you wish: `asprof --jfrsync +jdk.ObjectAllocationSample ...`
